File: cross-compatibility-testing/ts-generator/generator.ts
/**
 * TypeScript Test Data Generator for Cross-Compatibility Testing
 * 
 * Generates test data using TypeScript grimlock implementation
 * to be verified by Go grimlock implementation
 */

import * as fs from 'fs';
import * as path from 'path';
import grimlock from '../../typescript/grimlock';

interface TestData {
  keyPair: {
    privateKey: string;
    publicKey: string;
  };
  passcodeDerivation: {
    passcode: string;
    salt: string;
    params: {
      timeCost: number;
      memoryCost: number;
      parallelism: number;
    };
    derivedKey: string;
  };
  recoveryKeyDerivation: {
    recoveryKeyBytes: string;
    derivedKey: string;
  };
  privateKeyEncryption: {
    privateKey: string;
    encryptionKey: string;
    aad: string;
    encrypted: {
      iv: string;
      tag: string;
      ciphertext: string;
    };
  };
  messageEncryption: {
    payload: {
      userMessage: string;
      assistantResponse: string;
      optionalContext: Record<string, unknown>;
    };
    userKeyPair: {
      privateKey: string;
      publicKey: string;
    };
    context: {
      conversationId: string;
      messageId: string;
    };
    encrypted: {
      ephemeralPublicKey: string;
      iv: string;
      tag: string;
      ciphertext: string;
    };
  };
  ecdhTest: {
    alicePrivateKey: string;
    alicePublicKey: string;
    bobPrivateKey: string;
    bobPublicKey: string;
    sharedSecret: string;
  };
  recoveryKeyTest: {
    recoveryKey: {
      key: string;
      mnemonic: string;
    };
    privateKey: string;
    aad: string;
    encrypted: {
      iv: string;
      tag: string;
      ciphertext: string;
    };
  };
}

// Helper to convert Uint8Array to base64
function toBase64(data: Uint8Array): string {
  return Buffer.from(data).toString('base64');
}

// Helper to convert string to Uint8Array
function fromString(str: string): Uint8Array {
  return new TextEncoder().encode(str);
}

async function generateTestData(): Promise<void> {
  console.log('Generating TypeScript test data...');
  
  const testData: TestData = {} as TestData;

  // 1. Key Generation Test
  console.log('Generating key pair...');
  const keyPair = await grimlock.generateKeyPair();
  testData.keyPair = {
    privateKey: toBase64(keyPair.privateKey),
    publicKey: toBase64(keyPair.publicKey),
  };

  // 2. Passcode Key Derivation Test
  console.log('Deriving passcode key...');
  const passcode = 'MySecurePasscode123!';
  const salt = crypto.getRandomValues(new Uint8Array(32));
  const kdfParams = {
    salt,
    argon2Params: {
      timeCost: 4,
      memoryCost: 128 * 1024, // 128MB
      parallelism: 2,
    },
  };
  const derivedKey = await grimlock.derivePasscodeKey(passcode, kdfParams);
  testData.passcodeDerivation = {
    passcode,
    salt: toBase64(salt),
    params: {
      timeCost: kdfParams.argon2Params.timeCost,
      memoryCost: kdfParams.argon2Params.memoryCost,
      parallelism: kdfParams.argon2Params.parallelism,
    },
    derivedKey: toBase64(derivedKey),
  };

  // 3. Recovery Key Derivation Test
  console.log('Deriving recovery key...');
  const recoveryKey = await grimlock.generateRecoveryKey();
  const recoveryDerivedKey = await grimlock.deriveRecoveryKey(recoveryKey.raw);
  testData.recoveryKeyDerivation = {
    recoveryKeyBytes: toBase64(recoveryKey.raw),
    derivedKey: toBase64(recoveryDerivedKey),
  };

  // 4. Private Key Encryption Test
  console.log('Encrypting private key...');
  const privateKeyToEncrypt = keyPair.privateKey;
  const encryptionKey = derivedKey; // Use the derived key from passcode
  const aad = fromString('user@example.com');
  const encryptedPrivateKey = await grimlock.encryptPrivateKey(
    privateKeyToEncrypt,
    encryptionKey,
    aad
  );
  testData.privateKeyEncryption = {
    privateKey: toBase64(privateKeyToEncrypt),
    encryptionKey: toBase64(encryptionKey),
    aad: toBase64(aad),
    encrypted: {
      iv: toBase64(encryptedPrivateKey.iv),
      tag: toBase64(encryptedPrivateKey.tag),
      ciphertext: toBase64(encryptedPrivateKey.ciphertext),
    },
  };

  // 5. Message Encryption Test
  console.log('Encrypting message...');
  const userKeyPair = await grimlock.generateKeyPair();
  const payload = {
    userMessage: 'Hello, this is a test message!',
    assistantResponse: 'I understand. This is a response.',
    context: {
      timestamp: '2024-01-15T10:30:00Z',
      metadata: 'test',
    },
  };
  const context = {
    conversationId: 'conv-123',
    messageId: 'msg-456',
  };
  const encryptedMessage = await grimlock.encryptMessage(
    payload,
    userKeyPair.publicKey,
    context
  );
  testData.messageEncryption = {
    payload: {
      userMessage: payload.userMessage,
      assistantResponse: payload.assistantResponse,
      optionalContext: payload.context,
    },
    userKeyPair: {
      privateKey: toBase64(userKeyPair.privateKey),
      publicKey: toBase64(userKeyPair.publicKey),
    },
    context: {
      conversationId: context.conversationId,
      messageId: context.messageId,
    },
    encrypted: {
      ephemeralPublicKey: toBase64(encryptedMessage.ephemeralPublicKey),
      iv: toBase64(encryptedMessage.iv),
      tag: toBase64(encryptedMessage.tag),
      ciphertext: toBase64(encryptedMessage.ciphertext),
    },
  };

  // 6. ECDH Test
  console.log('Computing shared secret...');
  const aliceKeyPair = await grimlock.generateKeyPair();
  const bobKeyPair = await grimlock.generateKeyPair();
  
  // Import ECDH function
  const { computeSharedSecret } = await import('../../typescript/grimlock/versions/v1/ecdh');
  const sharedSecret = await computeSharedSecret(
    aliceKeyPair.privateKey,
    bobKeyPair.publicKey
  );
  testData.ecdhTest = {
    alicePrivateKey: toBase64(aliceKeyPair.privateKey),
    alicePublicKey: toBase64(aliceKeyPair.publicKey),
    bobPrivateKey: toBase64(bobKeyPair.privateKey),
    bobPublicKey: toBase64(bobKeyPair.publicKey),
    sharedSecret: toBase64(sharedSecret),
  };

  // 7. Recovery Key Test
  console.log('Testing recovery key encryption...');
  const testRecoveryKey = await grimlock.generateRecoveryKey();
  const privateKeyForRecovery = keyPair.privateKey;
  const aadForRecovery = fromString('recovery@example.com');
  const recoveryEncryptionKey = await grimlock.deriveRecoveryKey(testRecoveryKey.raw);
  const encryptedWithRecovery = await grimlock.encryptPrivateKey(
    privateKeyForRecovery,
    recoveryEncryptionKey,
    aadForRecovery
  );
  testData.recoveryKeyTest = {
    recoveryKey: {
      key: toBase64(testRecoveryKey.raw),
      mnemonic: testRecoveryKey.mnemonic || '',
    },
    privateKey: toBase64(privateKeyForRecovery),
    aad: toBase64(aadForRecovery),
    encrypted: {
      iv: toBase64(encryptedWithRecovery.iv),
      tag: toBase64(encryptedWithRecovery.tag),
      ciphertext: toBase64(encryptedWithRecovery.ciphertext),
    },
  };

  // Clean up sensitive data (best effort in JavaScript)
  keyPair.privateKey.fill(0);
  derivedKey.fill(0);
  recoveryDerivedKey.fill(0);
  encryptionKey.fill(0);
  userKeyPair.privateKey.fill(0);
  aliceKeyPair.privateKey.fill(0);
  bobKeyPair.privateKey.fill(0);
  sharedSecret.fill(0);
  recoveryEncryptionKey.fill(0);

  // Write to JSON file
  const outputPath = path.join(__dirname, '../test-data/ts-generated.json');
  const outputDir = path.dirname(outputPath);
  
  // Create directory if it doesn't exist
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }
  
  fs.writeFileSync(outputPath, JSON.stringify(testData, null, 2));
  console.log(`✅ Test data generated successfully: ${outputPath}`);
}

// Run the generator
generateTestData().catch((error) => {
  console.error('❌ Failed to generate test data:', error);
  process.exit(1);
});

File: cross-compatibility-testing/python-generator/generator.py
#!/usr/bin/env python3
"""Python Test Data Generator for Cross-Compatibility Testing

Generates test data using Python grimlock implementation
that can be verified by Go and TypeScript implementations.
"""

import base64
import json
import os
import sys
from pathlib import Path

# Add grimlock to path
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "python" / "grimlock"))

import grimlock
from grimlock.types.common import Argon2Params, KdfParams, MessageContext, MessagePayload


def main():
    """Generate test data."""
    test_data = {}

    g = grimlock.v1

    # 1. Key Generation Test
    print("Generating key pair...")
    key_pair = g.generate_key_pair()
    test_data["keyPair"] = {
        "privateKey": base64.b64encode(key_pair.private_key).decode("ascii"),
        "publicKey": base64.b64encode(key_pair.public_key).decode("ascii"),
    }

    # 2. Passcode Key Derivation Test
    print("Deriving passcode key...")
    passcode = "MySecurePasscode123!"
    salt = g.generate_salt()
    kdf_params = KdfParams(
        salt=salt,
        argon2_params=Argon2Params(
            time_cost=4,
            memory_cost=128 * 1024,  # 128MB
            parallelism=2,
        ),
    )
    derived_key = g.derive_passcode_key(passcode, kdf_params)
    test_data["passcodeDerivation"] = {
        "passcode": passcode,
        "salt": base64.b64encode(salt).decode("ascii"),
        "params": {
            "timeCost": kdf_params.argon2_params.time_cost,
            "memoryCost": kdf_params.argon2_params.memory_cost,
            "parallelism": kdf_params.argon2_params.parallelism,
        },
        "derivedKey": base64.b64encode(derived_key).decode("ascii"),
    }

    # 3. Recovery Key Derivation Test
    print("Deriving recovery key...")
    recovery_key = g.generate_recovery_key()
    recovery_derived_key = g.derive_recovery_key(recovery_key.raw)
    test_data["recoveryKeyDerivation"] = {
        "recoveryKeyBytes": base64.b64encode(recovery_key.raw).decode("ascii"),
        "derivedKey": base64.b64encode(recovery_derived_key).decode("ascii"),
    }

    # 4. Private Key Encryption Test
    print("Encrypting private key...")
    private_key_to_encrypt = key_pair.private_key
    encryption_key = derived_key  # Use the derived key from passcode
    aad = b"user@example.com"
    encrypted_private_key = g.encrypt_private_key(
        private_key_to_encrypt, encryption_key, aad
    )
    test_data["privateKeyEncryption"] = {
        "privateKey": base64.b64encode(private_key_to_encrypt).decode("ascii"),
        "encryptionKey": base64.b64encode(encryption_key).decode("ascii"),
        "aad": base64.b64encode(aad).decode("ascii"),
        "encrypted": {
            "iv": base64.b64encode(encrypted_private_key.iv).decode("ascii"),
            "tag": base64.b64encode(encrypted_private_key.tag).decode("ascii"),
            "ciphertext": base64.b64encode(encrypted_private_key.ciphertext).decode(
                "ascii"
            ),
        },
    }

    # 5. Message Encryption Test
    print("Encrypting message...")
    user_key_pair = g.generate_key_pair()
    payload = MessagePayload(
        user_message="Hello, this is a test message!",
        assistant_response="I understand. This is a response.",
        context={
            "timestamp": "2024-01-15T10:30:00Z",
            "metadata": "test",
        },
    )
    context = MessageContext(
        conversation_id="conv-123",
        message_id="msg-456",
    )
    encrypted_message = g.encrypt_message(payload, user_key_pair.public_key, context)
    test_data["messageEncryption"] = {
        "payload": {
            "userMessage": payload.user_message,
            "assistantResponse": payload.assistant_response,
            "optionalContext": payload.context or {},
        },
        "userKeyPair": {
            "privateKey": base64.b64encode(user_key_pair.private_key).decode("ascii"),
            "publicKey": base64.b64encode(user_key_pair.public_key).decode("ascii"),
        },
        "context": {
            "conversationId": context.conversation_id,
            "messageId": context.message_id,
        },
        "encrypted": {
            "ephemeralPublicKey": base64.b64encode(
                encrypted_message.ephemeral_public_key
            ).decode("ascii"),
            "iv": base64.b64encode(encrypted_message.iv).decode("ascii"),
            "tag": base64.b64encode(encrypted_message.tag).decode("ascii"),
            "ciphertext": base64.b64encode(encrypted_message.ciphertext).decode("ascii"),
        },
    }

    # 6. ECDH Test
    print("Computing shared secret...")
    alice_key_pair = g.generate_key_pair()
    bob_key_pair = g.generate_key_pair()
    shared_secret = g.compute_shared_secret(
        alice_key_pair.private_key, bob_key_pair.public_key
    )
    test_data["ecdhTest"] = {
        "alicePrivateKey": base64.b64encode(alice_key_pair.private_key).decode("ascii"),
        "alicePublicKey": base64.b64encode(alice_key_pair.public_key).decode("ascii"),
        "bobPrivateKey": base64.b64encode(bob_key_pair.private_key).decode("ascii"),
        "bobPublicKey": base64.b64encode(bob_key_pair.public_key).decode("ascii"),
        "sharedSecret": base64.b64encode(shared_secret).decode("ascii"),
    }

    # 7. Recovery Key Test
    print("Testing recovery key encryption...")
    test_recovery_key = g.generate_recovery_key()
    private_key_for_recovery = key_pair.private_key
    aad_for_recovery = b"recovery@example.com"
    recovery_encryption_key = g.derive_recovery_key(test_recovery_key.raw)
    encrypted_with_recovery = g.encrypt_private_key(
        private_key_for_recovery, recovery_encryption_key, aad_for_recovery
    )
    test_data["recoveryKeyTest"] = {
        "recoveryKey": {
            "key": base64.b64encode(test_recovery_key.raw).decode("ascii"),
            "mnemonic": test_recovery_key.mnemonic or "",
        },
        "privateKey": base64.b64encode(private_key_for_recovery).decode("ascii"),
        "aad": base64.b64encode(aad_for_recovery).decode("ascii"),
        "encrypted": {
            "iv": base64.b64encode(encrypted_with_recovery.iv).decode("ascii"),
            "tag": base64.b64encode(encrypted_with_recovery.tag).decode("ascii"),
            "ciphertext": base64.b64encode(encrypted_with_recovery.ciphertext).decode(
                "ascii"
            ),
        },
    }

    # Write to JSON file
    output_path = Path(__file__).parent.parent / "test-data" / "python-generated.json"
    output_path.parent.mkdir(parents=True, exist_ok=True)

    with open(output_path, "w") as f:
        json.dump(test_data, f, indent=2)

    print(f"✅ Test data generated successfully: {output_path}")


if __name__ == "__main__":
    main()
